Use `unitialized` for wildcard initializers

odersky opened the pull request:
Fixes #11225

smarter commented on docs/docs/reference/dropped-features/wildcard-init.md:
````diff
@@ -0,0 +1,18 @@
+```scala
+  var x: A = _
+```
+that was used to indicate an uninitialized field has been dropped.
+At its place there is a special value `notInitialized` in the `scala.compiletime` package. To get an uninitialized field, you now write
````
Minor bikeshed: since we call them uninitialized fields, why not call this method `uninitialized` too?

sjrd commented:
Whatever we do, if it touches the language and is not a critical issue, it should now wait for 3.1. We set that as a constraint to ourselves.

odersky commented:
I think it's such a small thing and such an obvious improvement that we should not hesitate to wait with it. We want to argue that Scala 3 is a lot simpler than Scala 2. This change is a clear win.

odersky commented:
Here's a reason for doing it now. The ProgFun2 MOOC contains two occurrences of the `= _` syntax.  Now is the last practical time to change the course.

sjrd commented:
> I think it's such a small thing and such an obvious improvement that we should not hesitate to wait with it. We want to argue that Scala 3 is a lot simpler than Scala 2. This change is a clear win.

A lot of things are "clear wins". With clear wins like this we can still be there at Easter. It's a language change, and it is not done as feedback about Scala 3 features (in fact it's not even a Scala 3 feature in the first place). According to the decisions we took so far, we should not do this, no matter how small of clear win it is.

> Here's a reason for doing it now. The ProgFun2 MOOC contains two occurrences of the = _ syntax. Now is the last practical time to change the course.

You said yourself that it's an obscure feature. How come ProgFun2 has two occurrences of this stuff? It seems to be in [`Signal.scala`](https://github.com/lampepfl/moocs/blob/master/courses/progfun2/calculator/src/main/scala/calculator/Signal.scala). There are ways to restructure this code not to need `= _` in the first place. I can send a PR if you want.

odersky commented:
@sjrd If you want to give it a go, I'd be interested to have a look at another `Signal` implementation that does not rely on uninitialized values. But if it should be a candidate for replacement, it has to be as short and as clear as the one given. Even then I am not sure I'll have the stomach to re-record.

odersky commented:
I also don't want to wait until Easter for 3.0. But adding this will not delay the release at all. We can merge now and be done with it.

sjrd commented:
> @sjrd If you want to give it a go, I'd be interested to have a look at another `Signal` implementation that does not rely on uninitialized values. But if it should be a candidate for replacement, it has to be as short and as clear as the one given. Even then I am not sure I'll have the stomach to re-record.

https://github.com/lampepfl/moocs/pull/592 (this is a link to a private repo)

sjrd commented:
Other relevant PR, this time public: https://github.com/lampepfl/dotty/pull/11238

dwijnand commented on docs/docs/reference/dropped-features/wildcard-init.md:
````diff
@@ -0,0 +1,18 @@
+```scala
+  var x: A = _
+```
+that was used to indicate an uninitialized field has been dropped.
````
```suggestion
that was used to indicate an uninitialized field, has been dropped.
```

Ideally I'd want to add a comma between "syntax" and "that" but the inline code block makes that awkward to communicate.  But, anyways, perhaps it's more clear that the syntax is dropped not the field.

dwijnand commented on library/src/scala/compiletime/package.scala:
```diff
@@ -18,6 +16,15 @@ package object compiletime {
+   *    var x: T = uninitialized
+   *
+   *  This signifies that the field is not initialized on its own (On the JVM
+   *  it is still bulk-initialized to a zero bitfield).
```
Too low-level an explanation?  Is it obvious or known enough that this mean `null`/`false`/`0`/`0L`/etc?

sjrd replied:
Also it's not just on the JVM. Scala.js and Scala Native have to replicate that behavior, obviously.

odersky commented:
@nicolasstucki Can you do a technical review for this?

nicolasstucki commented on docs/docs/reference/dropped-features/wildcard-init.md:
````diff
@@ -0,0 +1,18 @@
+  var x: A = _
+```
+that was used to indicate an uninitialized field, has been dropped.
+At its place there is a special value `uninitialized` in the `scala.compiletime` package. To get an uninitialized field, you now write
````
Why need an import for this language feature at all?

odersky replied:
Because it's a very special case. Normally, fields should be initialized. In the special cases where that's no the case and import is in order. That's not for novices, so knowing about compiletime is OK.

nicolasstucki commented on docs/docs/reference/dropped-features/wildcard-init.md:
````diff
@@ -0,0 +1,18 @@
+```scala
+import scala.compiletime.uninitialized
+
+var x: A = uninitialized
````
Currently, we also support

```scala
  var a = uninitialized[A]
```

This seems to be leaking implementation details.

odersky replied:
Right. We it's better to define it like this:
```scala
@compileTimeOnly("`uninitialized` can only be used as the right hand side of a mutable field definition")
  def uninitialized: Nothing = ???
```

nicolasstucki commented on library/src/scala/compiletime/package.scala:
```diff
@@ -18,6 +16,16 @@ package object compiletime {
+   *  as part of the bulk initialization of the object it belongs to, which assigns zero
+   *  values such as `null`, `0`, `0.0`, `false` to all object fields.
+   */
+  erased def uninitialized[T]: T = ???
```
Another error message that does not help anymore

```scala
35 |  var (a: Int, b: Int) = _
   |                         ^
   |                       Unbound placeholder parameter; incorrect use of _
```

```scala
-- Error: Foo.scala:34:22 ------------------------------------------------------
36 |  var (a: Int, b: Int) = uninitialized
   |                         ^^^^^^^^^^^^^
   |         method uninitialized is declared as erased, but is in fact used
```

sjrd replied:
Indeed. This is what I was already saying at https://github.com/lampepfl/dotty/issues/11225#issuecomment-771179394. `erased` is simply not the correct concept here.

nicolasstucki replied:
The current message is 
```scala
var (a: Int, b: Int) = uninitialized
                       ^^^^^^^^^^^^^ 
            `uninitialized` can only be used as the right hand side of a mutable field definition
```
but this looks like the `right hand side of a mutable field definition`

odersky replied:
Yes, borderline, but I think we can leave as is.

nicolasstucki commented on docs/docs/reference/syntax.md:
```diff
@@ -374,14 +374,12 @@ DefSig            ::=  id [DefTypeParamClause] DefParamClauses
 PatDef            ::=  ids [‘:’ Type] ‘=’ Expr
                     |  Pattern2 [‘:’ Type] ‘=’ Expr
-VarDef            ::=  PatDef
-                    |  ids ‘:’ Type ‘=’ ‘_’
```
We could avoid all the issues with inconsistent semantics and bad error messages if we define this as syntax.

```
VarDef            ::=  PatDef
                    |  ids ‘:’ Type ‘=’ ‘_’   // compat
                    |  ids ‘:’ Type ‘=’ ‘unimplemented’ 
```

Then simply desugar `var x: T = uninitialized ` into `var x: T = _` or any other internal representation we want. We could also desugar both into `var x: T = scala.runtime. uninitialized`. This would mean that all the checks and error messages for `var x: T = _` would also be used `var x: T = uninitialized`.

We could escape the syntax using ``var x: T = `uninitialized` `` if some app does use a `def uninitialized` already.

nicolasstucki replied:
This also has an advantage for the IDE and GitHub as we could highlight this `uninitialized` token.

nicolasstucki replied:
With this approach, we could still evolve the feature to somthing more complex in the future such as requiring an import and defining a method for it in the library. Those are the details that we should not rush to implement right now. It is only the syntax change that we care about.

odersky replied:
> Then simply desugar var x: T = uninitialized into var x: T = _ or any other internal representation we want.

That's actually what we do, in `PruneErasedDefs`. It's just that the official syntax does not mention `_ =` (just as it does not mention all the other 2.13 things we still support for now).

odersky replied:
No, I think that would be worse. We want a regular identifier that we can hover over in an IDE. I think @compileTimeOnly works very well for that. Otherwise you get into problems with soft keywords. It would be positively weird to _require_ a backticked `uninitialized` on the right hand side. That's just another obscure special case. We want to get away from these!

nicolasstucki commented on compiler/src/dotty/tools/dotc/transform/PruneErasedDefs.scala:
```diff
@@ -39,9 +44,16 @@ class PruneErasedDefs extends MiniPhase with SymTransformer { thisTransform =>
       cpy.ValDef(tree)(rhs = trivialErasedTree(tree))
-    else tree
+    else tree.rhs match
+      case rhs: TypeApply
```
If `uninitialized` then we should also expect aliasing to work
```scala
inline def uninitialised[A] = uninitialized[A]

var x: Int = uninitialised
```

Therefore we would also need to look into `Inline(_, _, _)`.

nicolasstucki reviewed (commented):
Seen how many things break with this approach it seems more reasonable to use well-defined syntax instead.

nicolasstucki reviewed (changes requested):
Seen how many things break with this approach it seems more reasonable to use well-defined syntax instead.

odersky commented:
@nicolasstucki I addressed the review comments. There's one remaining corner case: if `uninitialized` gets removed by inlining we do not flag it. That's just what `compileTimeOnly` and `erased` imply, so we should not doctor with it to 
change the behavior.

If you want to propose special syntax, you should say which one, and how to best explain it.

nicolasstucki commented on tests/neg/i11225.check:
```diff
@@ -0,0 +1,49 @@
+-- Error: tests/neg/i11225.scala:10:18 ---------------------------------------------------------------------------------
+10 |  var x6: Int = { uninitialized } // error
+   |                  ^^^^^^^^^^^^^
+   |                  `uninitialized` can only be used as the right hand side of a mutable field definition
```
This is on the right hand side of a mutable field definition. The message is inconsistent.

odersky replied:
The message says `as` not `on`. So I think it's precise.

nicolasstucki reviewed (changes requested):
There are still a couple of bugs and technical inconsistencies.

Given that the following case slipped through the test cases.
```scala
class Foo:
  var y0: Int = uninitialized // error
```
I would suggest improving test coverage by porting a few existing tests with `= _` to use `= uninitialized`.

odersky commented on tests/neg/i11225.scala:
```diff
@@ -0,0 +1,30 @@
+  inline def g(inline x: Int): Unit = ()
+  def f2 = g(uninitialized) // this one is ok since `uninitialized` is inlined away
+
+  var x7: Int = uni  // error
```
It's an error since we end up with `Inlined(uninitialized)`. I think that's OK. Ideally we would not allow the inline def of `uni` in the first place.

nicolasstucki replied:
Then it would be inconsistent with `var x4: Int = if false then uninitialized else 1 // ok, since inlined away` and `  def f2 = g(uninitialized) // this one is ok since uninitialized is inlined away`. We would have some special inlining rules that only apply `uninitialized` which will confuse everyone.

odersky replied:
I think we need  not overthink this. The intention is that you should not use `uninitialized`  except as the RHS of a field definition. For all practical purposes that's what this PR achieves. We don't need to split hairs beyond that,

odersky commented on tests/neg/i11225.scala:
```diff
@@ -0,0 +1,30 @@
+  val x1: Int = uninitialized // error
+  var x2: Int = if ??? then uninitialized else uninitialized // error // error
+  var x3: Int = if true then uninitialized else 1 // error
+  var x4: Int = if false then uninitialized else 1 // ok, since inlined away
```
It's OK with the latest commit.

nicolasstucki commented on compiler/src/dotty/tools/dotc/parsing/Parsers.scala:
```diff
@@ -3224,9 +3225,14 @@ object Parsers {
+              if sourceVersion.isAtLeast(`3.1`) then
+                deprecationWarning(
+                  em"""`= _` has been deprecated; use `= uninitialized` instead.
+                      |`uninitialized` needs to be imported from scala.compiletime.""", rhsOffset)
```
```suggestion
                      |`uninitialized` can be imported with `scala.compiletime.uninitialized`.""", rhsOffset)
```

nicolasstucki replied:
The tailing dot looked like an incomplete part the import: `scala.compiletime.`

djspiewak commented:
@odersky I realized today while pondering this (apparently too late) that this is more disruptive than I realized. The problem stems from the fact that this *cannot* be shimmed by a uniform inherited conditional compilation strategy. What I mean by that is abusing `src/main/scala-2`/`src/main/scala-3` directories and having some common parent type which encapsulates the differences between the languages.

In this case, the difference which must be encapsulated is *the initialization semantic in the constructor*. A concrete example is instructive: https://github.com/typelevel/cats-effect/blob/bde1abdb8aa80b1d0d765db33f7ba1825d529978/core/shared/src/main/scala/cats/effect/IOFiber.scala#L91-L92

The problem here is that these definitions cannot be extracted into the parent type without materially impacting their runtime semantics. Extraction into the parent is the standard trick here because it allows the conditional compilation to be isolated to just the parts of the implementation which must be fundamentally distinct between language versions. You can see a related example of this in that same type:

- JavaScript: https://github.com/typelevel/cats-effect/blob/bde1abdb8aa80b1d0d765db33f7ba1825d529978/core/js/src/main/scala/cats/effect/IOFiberPlatform.scala
- JVM: https://github.com/typelevel/cats-effect/blob/bde1abdb8aa80b1d0d765db33f7ba1825d529978/core/jvm/src/main/scala/cats/effect/IOFiberPlatform.scala

Unfortunately, in this case it is *impossible* to isolate the changes to just a parent type, exposing them via inheritance, without sacrificing some measurable elements of the current system. This is for two reasons. First, moving variables to a supertype changes their location in the object header, which has an impact on page faults and, in the worst case, can create added indirection (I'd have to actually test it, but I believe this *particular* case would indeed result in added indirection due to the need to retain some variables in the child class). This is a JVM limitation. Second, and in a sense, worse: the encoding of Scala's `protected` access modifier and uniform `var`s in the language means that it is impossible to obtain the same performance characteristics as a `private[this] var` from *any other access modifier*. All other access modifiers are encoded as methods which wrap instance fields, which actually (perhaps surprisingly!) has a measurable performance cost even after HotSpot has inlined the method body (which it will in most, but not all, cases).

And here we have the real crux of the problem: *no one* is using this syntax except in exceptionally performance-sensitive contexts, such as the file I linked. By definition, this feature is only meaningful for people who are sensitive to these kinds of performance degradations. But also by definition, it is impossible to *avoid* these kinds of degradations without doing one of two things:

- Giving up on cross-publishing for both Scala 2 and Scala 3 (the nuclear option that no one wants)
- Duplicating the *entire* file between Scala 2 and Scala 3 just for the sake of the new `uninitialized` syntax

The latter is the only option which will allow Cats Effect (and similar projects) to continue cross-publishing for Dotty, but unfortunately it also means we will be forced to needlessly duplicate the entire file in two source directories and manually keep them in sync. I invite you to peruse the file's contents further and judge for yourself how sustainable this is likely to be.

All in all, I come back to my original premise: this is simply not worth it. This change imposes a perhaps-surprisingly significant burden on an already-overstretched ecosystem. It would be one thing if this were an enormously beneficial language refinement (such as removing unbounded type projection) which carried manifest benefits across multiple domains, but it's just... not. This is a tiny bit of syntax that most people never see, and arguably isn't even that terrible when you *do* see it. The marginal gains are small, and the marginal losses are substantial and self-compounding.

Perhaps this is too late, but honestly and sincerely, I urge you to reconsider.

dwijnand commented:
A third option is back-porting this to 2.13 and 2.12.

LPTK commented:
As a first step, why not just add a shim in `src/main/scala-2/`?

```scala
def uninitialized[A]: A = null.asInstanceOf[A]
```

Then we can see if this actually results in any visible performance regression. My guess would be that it doesn't, though one can never be sure with the JVM.

djspiewak commented:
@LPTK I've actually tested the performance regression in this case previously. It is quite measurable.

Backporting is an option here. Also the fact that it's only deprecated and not removed (I missed that in my PR review) helps quite a bit (and avoids the worst case scenarios I described), though it does mean that fatal warnings are off the table for a while, and we will just have this same discussion again when 3.1 or 3.2 rolls around.

smarter commented:
> and we will just have this same discussion again when 3.1 or 3.2 rolls around.

There's a bunch of things which are currently deprecated under `-source 3.1` but since we'd like to be free to bump the minor version at any time, it's very likely we're going to move that to `-source future` and put off the deprecation until we can drop scala 2 support or scala 2 `-Xsource:3` supports the new form of things we want to deprecate, cf https://github.com/lampepfl/dotty/issues/11270

dwijnand commented:
Would a fourth option be making both compilers elide the bytecode emitted by `null.asInstanceOf[A]` or does that remove legitimate bytecode?

djspiewak commented:
> Would a fourth option be making both compilers elide the bytecode emitted by null.asInstanceOf[A] or does that remove legitimate bytecode?

It would mostly remove legitimate bytecode. Specifically, it would need to remove `x = null.asInstanceOf[A]`, but only in the initial declaration of the field (so, in the constructor).

sjrd commented:
That could alter the meaning of some (admittedly poorly designed) programs. It could be that something assigns a non-zero value to the field before we get to its initialization code (for example, from the super constructor, through a virtual method, or more simply from previous statements in the template). The explicitly assignment to `null` is supposed to erase that value. If you remove its bytecode, you remove that effect and change the meaning of the program.
